// @ai-sdk/azure/dist/index.mjs
// src/azure-openai-provider.ts
import {
  OpenAIChatLanguageModel,
  OpenAICompletionLanguageModel,
  OpenAIEmbeddingModel,
  OpenAIImageModel,
  OpenAIResponsesLanguageModel,
  OpenAISpeechModel,
  OpenAITranscriptionModel
} from "@ai-sdk/openai/internal";
import { DeepSeekChatLanguageModel } from "@ai-sdk/deepseek/internal";
import {
  InvalidArgumentError
} from "@ai-sdk/provider";
import {
  loadApiKey,
  loadSetting,
  normalizeHeaders,
  withUserAgentSuffix
} from "@ai-sdk/provider-utils";

// src/azure-openai-tools.ts
import {
  codeInterpreter,
  fileSearch,
  imageGeneration,
  webSearchPreview
} from "@ai-sdk/openai/internal";
var azureOpenaiTools = {
  codeInterpreter,
  fileSearch,
  imageGeneration,
  webSearchPreview
};

// src/version.ts
var VERSION = true ? "2.0.119" : "0.0.0-test";

// src/azure-openai-provider.ts
function createAzure(options = {}) {
  var _a;
  const tokenProvider = options.tokenProvider;
  if (options.apiKey && tokenProvider) {
    throw new InvalidArgumentError({
      argument: "apiKey/tokenProvider",
      message: "Both apiKey and tokenProvider were provided. Please use only one authentication method."
    });
  }
  const getHeaders = () => {
    const authHeaders = tokenProvider ? {} : {
      "api-key": loadApiKey({
        apiKey: options.apiKey,
        environmentVariableName: "AZURE_API_KEY",
        description: "Azure OpenAI"
      })
    };
    return withUserAgentSuffix(
      {
        ...authHeaders,
        ...options.headers
      },
      `ai-sdk/azure/${VERSION}`
    );
  };
  const fetch = tokenProvider ? async (input, init) => {
    var _a2;
    const headers = normalizeHeaders(init == null ? void 0 : init.headers);
    if (headers.authorization == null) {
      headers.authorization = `Bearer ${await tokenProvider()}`;
    }
    return ((_a2 = options.fetch) != null ? _a2 : globalThis.fetch)(input, {
      ...init,
      headers
    });
  } : options.fetch;
  const getResourceName = () => loadSetting({
    settingValue: options.resourceName,
    settingName: "resourceName",
    environmentVariableName: "AZURE_RESOURCE_NAME",
    description: "Azure OpenAI resource name"
  });
  const apiVersion = (_a = options.apiVersion) != null ? _a : "v1";
  const url = ({ path, modelId }) => {
    var _a2;
    const baseUrlPrefix = (_a2 = options.baseURL) != null ? _a2 : `https://${getResourceName()}.openai.azure.com/openai`;
    let fullUrl;
    if (options.useDeploymentBasedUrls) {
      fullUrl = new URL(`${baseUrlPrefix}/deployments/${modelId}${path}`);
    } else {
      fullUrl = new URL(`${baseUrlPrefix}/v1${path}`);
    }
    fullUrl.searchParams.set("api-version", apiVersion);
    return fullUrl.toString();
  };
  const createChatModel = (deploymentName) => new OpenAIChatLanguageModel(deploymentName, {
    provider: "azure.chat",
    url,
    headers: getHeaders,
    fetch
  });
  const createCompletionModel = (modelId) => new OpenAICompletionLanguageModel(modelId, {
    provider: "azure.completion",
    url,
    headers: getHeaders,
    fetch
  });
  const createEmbeddingModel = (modelId) => new OpenAIEmbeddingModel(modelId, {
    provider: "azure.embeddings",
    headers: getHeaders,
    url,
    fetch
  });
  const createResponsesModel = (modelId) => new OpenAIResponsesLanguageModel(modelId, {
    provider: "azure.responses",
    url,
    headers: getHeaders,
    fetch,
    fileIdPrefixes: ["assistant-"]
  });
  const createDeepSeekModel = (deploymentName) => new DeepSeekChatLanguageModel(deploymentName, {
    provider: "azure.deepseek",
    url,
    headers: getHeaders,
    fetch,
    supportsThinking: false,
    // json_object with thinking enabled makes Azure return the JSON in reasoning_content with empty content
    supportsStructuredOutputs: true
  });
  const createImageModel = (modelId) => new OpenAIImageModel(modelId, {
    provider: "azure.image",
    url,
    headers: getHeaders,
    fetch
  });
  const createTranscriptionModel = (modelId) => new OpenAITranscriptionModel(modelId, {
    provider: "azure.transcription",
    url,
    headers: getHeaders,
    fetch
  });
  const createSpeechModel = (modelId) => new OpenAISpeechModel(modelId, {
    provider: "azure.speech",
    url,
    headers: getHeaders,
    fetch
  });
  const provider = function(deploymentId) {
    if (new.target) {
      throw new Error(
        "The Azure OpenAI model function cannot be called with the new keyword."
      );
    }
    return createChatModel(deploymentId);
  };
  provider.languageModel = createChatModel;
  provider.chat = createChatModel;
  provider.deepseek = createDeepSeekModel;
  provider.completion = createCompletionModel;
  provider.embedding = createEmbeddingModel;
  provider.image = createImageModel;
  provider.imageModel = createImageModel;
  provider.textEmbedding = createEmbeddingModel;
  provider.textEmbeddingModel = createEmbeddingModel;
  provider.responses = createResponsesModel;
  provider.transcription = createTranscriptionModel;
  provider.speech = createSpeechModel;
  provider.tools = azureOpenaiTools;
  return provider;
}
var azure = createAzure();
export {
  VERSION,
  azure,
  createAzure
};
