// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/clipboard.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextClipboard = void 0;
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
const FlowLogger_js_1 = require("../flowlogger/FlowLogger.js");
let ContextClipboard = (() => {
    let _instanceExtraInitializers = [];
    let _writeText_decorators;
    let _readText_decorators;
    let _clear_decorators;
    let _paste_decorators;
    let _copy_decorators;
    let _cut_decorators;
    return class ContextClipboard {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _writeText_decorators = [FlowLogger_js_1.FlowLogger.wrapWithLogging({ eventType: "ClipboardWriteText" })];
            _readText_decorators = [FlowLogger_js_1.FlowLogger.wrapWithLogging({ eventType: "ClipboardReadText" })];
            _clear_decorators = [FlowLogger_js_1.FlowLogger.wrapWithLogging({ eventType: "ClipboardClear" })];
            _paste_decorators = [FlowLogger_js_1.FlowLogger.wrapWithLogging({ eventType: "ClipboardPaste" })];
            _copy_decorators = [FlowLogger_js_1.FlowLogger.wrapWithLogging({ eventType: "ClipboardCopy" })];
            _cut_decorators = [FlowLogger_js_1.FlowLogger.wrapWithLogging({ eventType: "ClipboardCut" })];
            __esDecorate(this, null, _writeText_decorators, { kind: "method", name: "writeText", static: false, private: false, access: { has: obj => "writeText" in obj, get: obj => obj.writeText }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _readText_decorators, { kind: "method", name: "readText", static: false, private: false, access: { has: obj => "readText" in obj, get: obj => obj.readText }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _clear_decorators, { kind: "method", name: "clear", static: false, private: false, access: { has: obj => "clear" in obj, get: obj => obj.clear }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _paste_decorators, { kind: "method", name: "paste", static: false, private: false, access: { has: obj => "paste" in obj, get: obj => obj.paste }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _copy_decorators, { kind: "method", name: "copy", static: false, private: false, access: { has: obj => "copy" in obj, get: obj => obj.copy }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _cut_decorators, { kind: "method", name: "cut", static: false, private: false, access: { has: obj => "cut" in obj, get: obj => obj.cut }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        params = __runInitializers(this, _instanceExtraInitializers);
        constructor(params) {
            this.params = params;
        }
        async writeText(text, options) {
            await this.writeTextInternal(text, options);
        }
        async writeTextInternal(text, options) {
            const page = await this.resolvePage(options?.page);
            await this.ensurePageFocused(page);
            await this.grantClipboardPermissions(page);
            await page.sendCDP("Runtime.enable").catch(() => { });
            const response = await page.sendCDP("Runtime.evaluate", {
                expression: `navigator.clipboard.writeText(${JSON.stringify(text)})`,
                awaitPromise: true,
                returnByValue: true,
                userGesture: true,
            });
            this.throwIfEvaluationFailed("write clipboard text", response);
        }
        async readText(options) {
            const page = await this.resolvePage(options?.page);
            await this.ensurePageFocused(page);
            await this.grantClipboardPermissions(page);
            await page.sendCDP("Runtime.enable").catch(() => { });
            const response = await page.sendCDP("Runtime.evaluate", {
                expression: "navigator.clipboard.readText()",
                awaitPromise: true,
                returnByValue: true,
                userGesture: true,
            });
            this.throwIfEvaluationFailed("read clipboard text", response);
            return String(response.result?.value ?? "");
        }
        async clear(options) {
            await this.writeTextInternal("", options);
        }
        async paste(options) {
            const page = await this.resolvePage(options?.page);
            await this.ensurePageFocused(page);
            await page.keyPress(options?.shortcut ?? "ControlOrMeta+V");
        }
        async copy(options) {
            const page = await this.resolvePage(options?.page);
            await this.ensurePageFocused(page);
            await page.keyPress("ControlOrMeta+C");
        }
        async cut(options) {
            const page = await this.resolvePage(options?.page);
            await this.ensurePageFocused(page);
            await page.keyPress("ControlOrMeta+X");
        }
        async resolvePage(page) {
            return await this.params.resolvePage(page);
        }
        async grantClipboardPermissions(page) {
            const origin = this.originForPage(page);
            if (!origin)
                return;
            await this.params.context.conn
                .send("Browser.grantPermissions", {
                origin,
                permissions: ["clipboardReadWrite", "clipboardSanitizedWrite"],
            })
                .catch(() => { });
        }
        async ensurePageFocused(page) {
            this.params.context.setActivePage(page);
            await page.sendCDP("Page.bringToFront").catch(() => { });
            await page.sendCDP("Runtime.enable").catch(() => { });
            await page
                .sendCDP("Runtime.evaluate", {
                expression: "window.focus()",
                awaitPromise: true,
                userGesture: true,
            })
                .catch(() => { });
        }
        originForPage(page) {
            try {
                const parsed = new URL(page.url());
                if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
                    return undefined;
                }
                return parsed.origin;
            }
            catch {
                return undefined;
            }
        }
        throwIfEvaluationFailed(operation, response) {
            if (!response.exceptionDetails)
                return;
            const message = response.exceptionDetails.exception?.description ??
                response.exceptionDetails.text ??
                "Runtime.evaluate failed";
            throw new sdkErrors_js_1.StagehandEvalError(`Failed to ${operation}: ${message}`);
        }
    };
})();
exports.ContextClipboard = ContextClipboard;
