// @browserbasehq/stagehand/dist/cjs/lib/v3/verifier/evidenceNormalization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REDACTED_INLINE_IMAGE = void 0;
exports.collectInlineImagePayloads = collectInlineImagePayloads;
exports.redactInlineImagePayloads = redactInlineImagePayloads;
exports.mergeAgentEvidence = mergeAgentEvidence;
exports.buildAgentEvidenceFromStepFinished = buildAgentEvidenceFromStepFinished;
exports.REDACTED_INLINE_IMAGE = "[redacted inline image payload]";
const INLINE_IMAGE_KEYS = new Set(["screenshotBase64"]);
function shouldRedactBase64Key(key, actionName) {
    return (INLINE_IMAGE_KEYS.has(key) ||
        (actionName === "screenshot" && key === "base64"));
}
function collectInlineImagePayloads(value, actionName, out = []) {
    if (!value || typeof value !== "object")
        return out;
    if (Buffer.isBuffer(value))
        return out;
    if (Array.isArray(value)) {
        for (const item of value) {
            collectInlineImagePayloads(item, actionName, out);
        }
        return out;
    }
    for (const [key, nested] of Object.entries(value)) {
        if (shouldRedactBase64Key(key, actionName) && typeof nested === "string") {
            out.push(nested);
            continue;
        }
        collectInlineImagePayloads(nested, actionName, out);
    }
    return out;
}
function redactInlineImagePayloads(value, actionName) {
    if (!value || typeof value !== "object")
        return value;
    if (Buffer.isBuffer(value))
        return value;
    if (Array.isArray(value)) {
        return value.map((item) => redactInlineImagePayloads(item, actionName));
    }
    const out = {};
    for (const [key, nested] of Object.entries(value)) {
        out[key] =
            shouldRedactBase64Key(key, actionName) && typeof nested === "string"
                ? exports.REDACTED_INLINE_IMAGE
                : redactInlineImagePayloads(nested, actionName);
    }
    return out;
}
function mergeAgentEvidence(...parts) {
    return {
        modalities: parts.flatMap((p) => p?.modalities ?? []),
    };
}
function buildAgentEvidenceFromStepFinished(event) {
    const modalities = [];
    if (event.reasoning) {
        modalities.push({ type: "text", content: event.reasoning });
    }
    const result = event.toolOutput.result;
    if (result === undefined || result === null) {
        return { modalities };
    }
    if (typeof result === "string") {
        modalities.push({ type: "text", content: result });
    }
    else if (typeof result === "number" ||
        typeof result === "boolean" ||
        typeof result === "bigint") {
        modalities.push({ type: "text", content: String(result) });
    }
    else if (Buffer.isBuffer(result)) {
        modalities.push({
            type: "image",
            bytes: result,
            mediaType: "image/png",
        });
    }
    else if (typeof result === "object") {
        for (const imageBase64 of collectInlineImagePayloads(result, event.actionName)) {
            try {
                modalities.push({
                    type: "image",
                    bytes: Buffer.from(imageBase64, "base64"),
                    mediaType: "image/png",
                });
            }
            catch {
                // Malformed base64; skip the image and keep the JSON modality.
            }
        }
        modalities.push({
            type: "json",
            content: redactInlineImagePayloads(result, event.actionName),
        });
    }
    return { modalities };
}
