// @browserbasehq/stagehand/dist/cjs/lib/v3/verifier/trajectory.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeRubric = normalizeRubric;
exports.loadTrajectoryFromDisk = loadTrajectoryFromDisk;
exports.nextResultFilename = nextResultFilename;
exports.shouldPersistTrajectory = shouldPersistTrajectory;
exports.writeTrajectoryDir = writeTrajectoryDir;
const promises_1 = __importDefault(require("node:fs/promises"));
const node_path_1 = __importDefault(require("node:path"));
const evidenceNormalization_js_1 = require("./evidenceNormalization.js");
/**
 * Convert dataset or generated rubric JSON into the public Stagehand shape.
 * Snake-case dataset fields are accepted here so serialized quirks do not leak
 * into the canonical rubric type.
 */
function normalizeRubric(rubric) {
    if (rubric == null)
        return undefined;
    if (typeof rubric !== "object") {
        throw new TypeError("Rubric must be an object");
    }
    const rawRubric = rubric;
    if (!Array.isArray(rawRubric.items)) {
        throw new TypeError("Rubric is missing an items array");
    }
    return {
        items: rawRubric.items.map((item) => {
            const criterion = normalizeRequiredString(item.criterion, "criterion");
            const description = normalizeRequiredString(item.description, "description");
            const maxPoints = normalizeMaxPoints(item);
            if (typeof maxPoints !== "number" || !Number.isFinite(maxPoints)) {
                throw new TypeError(`Rubric criterion "${criterion}" is missing a numeric maxPoints value`);
            }
            return {
                criterion,
                description,
                maxPoints,
                ...(typeof item.condition === "string" && {
                    condition: item.condition,
                }),
            };
        }),
    };
}
function normalizeRequiredString(value, fieldName) {
    if (typeof value === "string" && value.length) {
        return value;
    }
    throw new TypeError(`Rubric criterion is missing a ${fieldName} value`);
}
function normalizeMaxPoints(item) {
    return item.maxPoints ?? item.max_points;
}
function normalizeResultLabel(label) {
    return (label ?? `rescore-${new Date().toISOString()}`).replace(/[^A-Za-z0-9._-]/g, "_");
}
// ─────────────────────────────────────────────────────────────────────────────
// On-disk loader
// ─────────────────────────────────────────────────────────────────────────────
/**
 * Hydrate a Trajectory from the on-disk directory layout written by
 * TrajectoryRecorder.persist(). Used by the offline re-scoring CLI (`bench
 * verify`) and by any consumer that wants to feed a saved trajectory back
 * into V3Evaluator.verify() without running an agent.
 *
 * Reverses the recorder's serialization tweaks:
 *   - `probeEvidence.screenshotPath` → read file into `probeEvidence.screenshot`.
 *   - Image modalities in `agentEvidence.modalities` carry `imagePath` on disk
 *     (relative to the trajectory dir). Legacy `bytesBase64` is also accepted
 *     for trajectories written before the externalization change.
 *
 * @param dir absolute or cwd-relative path to a `<run-id>/<task-id>/` directory.
 */
async function loadTrajectoryFromDisk(dir) {
    const fs = await Promise.resolve().then(() => __importStar(require("node:fs/promises")));
    const path = await Promise.resolve().then(() => __importStar(require("node:path")));
    const trajectoryDir = path.resolve(dir);
    const trajectoryPath = path.join(trajectoryDir, "trajectory.json");
    const raw = await fs.readFile(trajectoryPath, "utf8");
    const parsed = JSON.parse(raw);
    const resolveWithinTrajectoryDir = (candidate, fieldName = "screenshotPath") => {
        const resolved = path.resolve(trajectoryDir, candidate);
        const relative = path.relative(trajectoryDir, resolved);
        const outside = relative === ".." ||
            relative.startsWith(`..${path.sep}`) ||
            path.isAbsolute(relative);
        if (outside) {
            throw new Error(`Trajectory ${fieldName} escapes trajectory directory: ${candidate}`);
        }
        return resolved;
    };
    const hydrateProbeScreenshot = async (probe) => {
        if (probe?.screenshotPath && !probe.screenshot) {
            const resolved = resolveWithinTrajectoryDir(probe.screenshotPath);
            try {
                probe.screenshot = await fs.readFile(resolved);
            }
            catch {
                // Missing screenshot file: leave probe.screenshot unset. The verifier's
                // evidence_insufficient path will handle it.
            }
        }
    };
    for (const step of parsed.steps) {
        // Rehydrate tier-2 probe screenshot from its on-disk file reference.
        await hydrateProbeScreenshot(step.probeEvidence);
        // Rehydrate tier-1 image modalities. New shape: resolve imagePath → Buffer.
        // Legacy shape: decode bytesBase64 → Buffer.
        if (step.agentEvidence?.modalities) {
            const resolved = [];
            for (const m of step.agentEvidence.modalities) {
                if (m.type !== "image") {
                    resolved.push(m);
                    continue;
                }
                const raw = m;
                if (typeof raw.imagePath === "string") {
                    const imagePath = resolveWithinTrajectoryDir(raw.imagePath, "imagePath");
                    try {
                        const bytes = await fs.readFile(imagePath);
                        resolved.push({ type: "image", bytes, mediaType: m.mediaType });
                    }
                    catch {
                        // Missing tier-1 image: drop this modality. evidence_insufficient
                        // handles the downstream effect.
                    }
                }
                else if (typeof raw.bytesBase64 === "string") {
                    resolved.push({
                        type: "image",
                        bytes: Buffer.from(raw.bytesBase64, "base64"),
                        mediaType: m.mediaType,
                    });
                }
                else {
                    resolved.push(m);
                }
            }
            step.agentEvidence.modalities = resolved;
        }
    }
    await hydrateProbeScreenshot(parsed.finalObservation);
    return parsed;
}
/**
 * Build a `result*.json` filename for persisted evaluator output.
 *
 * Convention: the live run writes `result.json`; offline re-score attempts use
 * a label-based name (e.g., `result_rescore-2026-05-11.json`) so they coexist
 * without collisions and remain easy to diff.
 */
function nextResultFilename(label) {
    return `result_${normalizeResultLabel(label)}.json`;
}
/**
 * Default persistence policy: explicit override, then env, then "on unless CI".
 */
function shouldPersistTrajectory(override) {
    if (override !== undefined)
        return override;
    const env = process.env.VERIFIER_PERSIST_TRAJECTORIES?.toLowerCase();
    if (env === "1" || env === "true")
        return true;
    if (env === "0" || env === "false")
        return false;
    return !process.env.CI;
}
/**
 * Write the on-disk trajectory layout under `dir`:
 *
 *   <dir>/
 *     ├── task_data.json
 *     ├── trajectory.json    (screenshots referenced by path)
 *     ├── screenshots/
 *     │   ├── probe/<N>.png
 *     │   └── agent/<N>[_M].png
 *     ├── scores/            (empty; populated separately)
 *     └── core.log
 *
 * Image bytes are externalized to PNG files; the in-memory Trajectory is left
 * untouched so callers can keep using it after persistence.
 */
async function writeTrajectoryDir(dir, trajectory) {
    await promises_1.default.mkdir(dir, { recursive: true });
    await promises_1.default.mkdir(node_path_1.default.join(dir, "screenshots", "probe"), { recursive: true });
    await promises_1.default.mkdir(node_path_1.default.join(dir, "screenshots", "agent"), { recursive: true });
    const serializableSteps = [];
    // A single post-turn probe is fanned across every step of a multi-tool turn,
    // and a single agent screenshot is shared across every action a CUA provider
    // chose from it, so the same Buffer is shared by reference. Dedupe by
    // identity: write the PNG once and point every sharing step at the same file.
    const probePathByBuffer = new Map();
    const agentPathByBuffer = new Map();
    for (const [i, step] of trajectory.steps.entries()) {
        const probe = { ...step.probeEvidence };
        if (probe.screenshot) {
            let relPath = probePathByBuffer.get(probe.screenshot);
            if (!relPath) {
                relPath = `screenshots/probe/${i + 1}.png`;
                await promises_1.default.writeFile(node_path_1.default.join(dir, relPath), probe.screenshot);
                probePathByBuffer.set(probe.screenshot, relPath);
            }
            probe.screenshotPath = relPath;
            delete probe.screenshot;
        }
        const imageModalities = step.agentEvidence.modalities.filter((m) => m.type === "image");
        const multipleImages = imageModalities.length > 1;
        let imageSeq = 0;
        const modalities = [];
        for (const m of step.agentEvidence.modalities) {
            if (m.type !== "image") {
                modalities.push(m.type === "json"
                    ? {
                        ...m,
                        content: (0, evidenceNormalization_js_1.redactInlineImagePayloads)(m.content, step.actionName),
                    }
                    : m);
                continue;
            }
            let relPath = agentPathByBuffer.get(m.bytes);
            if (!relPath) {
                const suffix = multipleImages ? `_${imageSeq}` : "";
                relPath = `screenshots/agent/${i + 1}${suffix}.png`;
                await promises_1.default.writeFile(node_path_1.default.join(dir, relPath), m.bytes);
                agentPathByBuffer.set(m.bytes, relPath);
            }
            modalities.push({
                type: "image",
                imagePath: relPath,
                mediaType: m.mediaType,
            });
            imageSeq += 1;
        }
        serializableSteps.push({
            ...step,
            probeEvidence: probe,
            agentEvidence: { modalities },
            toolOutput: {
                ...step.toolOutput,
                result: (0, evidenceNormalization_js_1.redactInlineImagePayloads)(step.toolOutput.result, step.actionName),
            },
        });
    }
    const finalObservation = trajectory.finalObservation === undefined
        ? undefined
        : { ...trajectory.finalObservation };
    if (finalObservation?.screenshot) {
        const relPath = "screenshots/probe/final.png";
        await promises_1.default.writeFile(node_path_1.default.join(dir, relPath), finalObservation.screenshot);
        finalObservation.screenshotPath = relPath;
        delete finalObservation.screenshot;
    }
    // Image modalities carry imagePath instead of raw bytes on disk; cast
    // through unknown rather than widen Trajectory's type contract.
    const serialized = {
        ...trajectory,
        steps: serializableSteps,
        ...(finalObservation ? { finalObservation } : {}),
    };
    await promises_1.default.writeFile(node_path_1.default.join(dir, "trajectory.json"), JSON.stringify(serialized, null, 2));
    await promises_1.default.writeFile(node_path_1.default.join(dir, "task_data.json"), JSON.stringify({
        task: trajectory.task,
        status: trajectory.status,
        finalAnswer: trajectory.finalAnswer ?? null,
    }, null, 2));
    await promises_1.default.mkdir(node_path_1.default.join(dir, "scores"), { recursive: true });
    await promises_1.default.writeFile(node_path_1.default.join(dir, "core.log"), coreLog(trajectory));
}
function coreLog(trajectory) {
    return (trajectory.steps
        .map((step, i) => JSON.stringify({
        step: i,
        action: step.actionName,
        url: step.probeEvidence.url ?? null,
        ok: step.toolOutput.ok,
        reasoning: step.reasoning || undefined,
    }))
        .join("\n") + "\n");
}
