// @browserbasehq/stagehand/dist/cjs/lib/v3/launch/local.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.launchLocalChrome = launchLocalChrome;
const chrome_launcher_1 = require("chrome-launcher");
const ws_1 = __importDefault(require("ws"));
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
const STAGEHAND_DEFAULT_FLAGS = [
    "--remote-allow-origins=*",
    "--no-first-run",
    "--no-default-browser-check",
    "--disable-dev-shm-usage",
    "--site-per-process",
    "--enable-features=WebMCPTesting,DevToolsWebMCPSupport",
];
async function launchLocalChrome(opts) {
    const connectTimeoutMs = opts.connectTimeoutMs ?? 15_000;
    const deadlineMs = Date.now() + connectTimeoutMs;
    const connectionPollInterval = 250;
    const maxConnectionRetries = Math.max(1, Math.ceil(connectTimeoutMs / connectionPollInterval));
    const headless = opts.headless ?? false;
    const ignoredDefaults = Array.isArray(opts.ignoreDefaultArgs)
        ? opts.ignoreDefaultArgs
        : null;
    const stagehandDefaultFlags = opts.ignoreDefaultArgs === true
        ? []
        : ignoredDefaults
            ? STAGEHAND_DEFAULT_FLAGS.filter((f) => !ignoredDefaults.includes(f))
            : [...STAGEHAND_DEFAULT_FLAGS];
    const chromeFlags = [
        ...stagehandDefaultFlags,
        headless ? "--headless=new" : undefined,
        opts.devtools ? "--auto-open-devtools-for-tabs" : undefined,
        opts.locale ? `--lang=${opts.locale}` : undefined,
        opts.viewport?.width && opts.viewport?.height
            ? `--window-size=${opts.viewport.width},${opts.viewport.height + 87}`
            : undefined,
        typeof opts.deviceScaleFactor === "number"
            ? `--force-device-scale-factor=${Math.max(0.1, opts.deviceScaleFactor)}`
            : undefined,
        opts.hasTouch ? "--touch-events=enabled" : undefined,
        opts.ignoreHTTPSErrors ? "--ignore-certificate-errors" : undefined,
        opts.proxy?.server ? `--proxy-server=${opts.proxy.server}` : undefined,
        opts.proxy?.bypass ? `--proxy-bypass-list=${opts.proxy.bypass}` : undefined,
        ...(opts.args ?? []),
    ].filter((f) => typeof f === "string");
    // Handle ignoreDefaultArgs: selectively remove chrome-launcher's built-in
    // defaults while keeping Stagehand's own flags (already in chromeFlags).
    let ignoreDefaultFlags = false;
    if (opts.ignoreDefaultArgs === true) {
        ignoreDefaultFlags = true;
    }
    else if (Array.isArray(opts.ignoreDefaultArgs) &&
        opts.ignoreDefaultArgs.length > 0) {
        // Tell chrome-launcher to skip ALL its defaults, then re-add the ones
        // the user did NOT ask to exclude.
        ignoreDefaultFlags = true;
        const excludeArgs = opts.ignoreDefaultArgs;
        const clDefaults = chrome_launcher_1.Launcher.defaultFlags?.() ?? [];
        const kept = clDefaults.filter((f) => !excludeArgs.includes(f));
        chromeFlags.unshift(...kept);
    }
    const chrome = await (0, chrome_launcher_1.launch)({
        chromePath: opts.executablePath,
        chromeFlags,
        port: opts.port,
        userDataDir: opts.userDataDir,
        handleSIGINT: opts.handleSIGINT,
        ignoreDefaultFlags,
        connectionPollInterval,
        maxConnectionRetries,
    });
    const ws = await waitForWebSocketDebuggerUrl(chrome.port, deadlineMs);
    await waitForWebSocketReady(ws, deadlineMs);
    return { ws, chrome };
}
async function waitForWebSocketDebuggerUrl(port, deadlineMs) {
    let lastErrMsg = "";
    while (Date.now() < deadlineMs) {
        try {
            const resp = await fetch(`http://127.0.0.1:${port}/json/version`);
            if (resp.ok) {
                const json = (await resp.json());
                const url = json
                    .webSocketDebuggerUrl;
                if (typeof url === "string")
                    return url;
            }
            else {
                lastErrMsg = `${resp.status} ${resp.statusText}`;
            }
        }
        catch (err) {
            lastErrMsg = err instanceof Error ? err.message : String(err);
        }
        await new Promise((r) => setTimeout(r, 250));
    }
    throw new sdkErrors_js_1.ConnectionTimeoutError(`Timed out waiting for /json/version on port ${port} ${lastErrMsg ? ` (last error: ${lastErrMsg})` : ""}`);
}
async function waitForWebSocketReady(wsUrl, deadlineMs) {
    let lastErrMsg = "";
    while (Date.now() < deadlineMs) {
        const remainingMs = Math.max(200, deadlineMs - Date.now());
        try {
            await probeWebSocket(wsUrl, Math.min(2_000, remainingMs));
            return;
        }
        catch (error) {
            lastErrMsg = error instanceof Error ? error.message : String(error);
            await new Promise((r) => setTimeout(r, 100));
        }
    }
    throw new sdkErrors_js_1.ConnectionTimeoutError(`Timed out waiting for CDP websocket to accept connections at ${wsUrl}${lastErrMsg ? ` (last error: ${lastErrMsg})` : ""}`);
}
function probeWebSocket(wsUrl, timeoutMs) {
    return new Promise((resolve, reject) => {
        const ws = new ws_1.default(wsUrl);
        let settled = false;
        const finish = (error) => {
            if (settled)
                return;
            settled = true;
            clearTimeout(timer);
            try {
                ws.terminate();
            }
            catch {
                // best-effort cleanup
            }
            if (error) {
                reject(error);
                return;
            }
            resolve();
        };
        const timer = setTimeout(() => {
            finish(new Error(`websocket probe timeout after ${timeoutMs}ms`));
        }, timeoutMs);
        ws.once("open", () => finish());
        ws.once("error", (error) => finish(error));
    });
}
