// @browserbasehq/stagehand/dist/cjs/lib/modelUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AUTO_MODEL_NAME = void 0;
exports.isAutoModel = isAutoModel;
exports.extractModelName = extractModelName;
exports.splitModelName = splitModelName;
exports.resolveModel = resolveModel;
const agent_js_1 = require("./v3/types/public/agent.js");
/**
 * Sentinel model name that delegates model selection to the Stagehand API.
 * Only valid when running through the API (env: "BROWSERBASE" with
 * disableAPI: false); no local LLM client or provider API key is involved.
 */
exports.AUTO_MODEL_NAME = "auto";
function isAutoModel(model) {
    return extractModelName(model) === exports.AUTO_MODEL_NAME;
}
//useful when resolving a model from string or object formats we accept
function extractModelName(model) {
    if (!model)
        return undefined;
    return typeof model === "string" ? model : model.modelName;
}
function splitModelName(model) {
    const firstSlashIndex = model.indexOf("/");
    const provider = model.substring(0, firstSlashIndex);
    const modelName = model.substring(firstSlashIndex + 1);
    return { provider, modelName };
}
function resolveModel(model) {
    const modelString = extractModelName(model);
    const clientOptions = typeof model === "string"
        ? {}
        : (() => {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { modelName: _, ...rest } = model;
            return rest;
        })();
    // Check if provider is explicitly set in clientOptions
    const hasExplicitProvider = clientOptions.provider !== undefined;
    // If provider is explicitly set, don't split the model name - pass it through as-is
    let provider;
    let parsedModelName;
    if (hasExplicitProvider) {
        provider = clientOptions.provider;
        parsedModelName = modelString; // Keep the full model name
    }
    else {
        // Parse the model string normally
        const split = splitModelName(modelString);
        provider = split.provider;
        parsedModelName = split.modelName;
    }
    // Check if it's a CUA model
    const isCua = hasExplicitProvider ||
        agent_js_1.AVAILABLE_CUA_MODELS.includes(modelString);
    return {
        provider,
        modelName: parsedModelName,
        clientOptions,
        isCua,
    };
}
